### PR TITLE
[jit] Add more SSE2 opcodes. Enable SSE2.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -7820,6 +7965,222 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			LLVMValueRef value = lhs;
+			LLVMValueRef index = rhs;
+			LLVMValueRef phi_values [64 + 1];
+			IntrinsicId id;
+			int ncases;
+
+			// FIXME: Optimize const index case
+
+			switch (ins->opcode) {
+			case OP_SSE2_PSRAW_IMM:
```
Comment (EgorBo) [inline src/mono/mono/mini/mini-llvm.c]: it stands for `ShiftRightArithmetic` which is `_mm_srai_epi16` which works with non-constant arguments normally https://godbolt.org/z/xWq6za why do we need to emit a huge switch for it?
Also I'd just throw an exception for non-cns arguments where intrinsics expect cns.

Comment (EgorBo) [reply src/mono/mono/mini/mini-llvm.c]: E.g. https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/simd-intrinsics-netcore.c#L881-L885

Comment (EgorBo) [reply src/mono/mono/mini/mini-llvm.c]: Here is what `ShiftRightArithmetic` emits now: https://gist.github.com/EgorBo/ea3f749413d573f5f1d4a522ff7cf841

Comment (imhameed) [reply src/mono/mono/mini/mini-llvm.c]: Looks like coreclr will just emit the non-immediate form of psraw when the shift amount can't be folded into a constant: https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4CS5AThaRBQMYT4AaAHAGwBudFlwES5KrQbBmrDhGFoRAZlwJsAYWwBvdNgO41ANWrtgAe0ZIEvADwyk/AHzYAsgApT5qzfuOXbAAbAAsIABpsACMAT2BqbDAoYABKXX1DTLgAdmwAZQhqBHw8kIoAM2BCCgBzEOAAQWZgEJpKdg9QiMTklKVMgF90AaA==

I agree that the switch doesn't seem necessary; we could just rely on LLVM to optimize the IR and select either the immediate or non-immediate form of these instructions as appropriate. https://godbolt.org/z/Zxagk4

File: src/mono/mono/mini/mini-llvm.c [hunk 2]
```diff
@@ -8026,64 +8026,27 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
-			phi_values [ncases] = LLVMConstNull (LLVMTypeOf (value));
-			LLVMBuildBr (builder, cbb);
-
-			LLVMPositionBuilderAtEnd (builder, cbb);
-			values [ins->dreg] = LLVMBuildPhi (builder, LLVMTypeOf (phi_values [0]), "");
-			LLVMAddIncoming (values [ins->dreg], phi_values, bbs, ncases + 1);
 
-			ctx->bblocks [bb->block_num].end_bblock = cbb;
+			LLVMTypeRef t = LLVMTypeOf (value);
+			LLVMValueRef index_vect = LLVMBuildInsertElement (builder, LLVMConstNull (t), convert (ctx, index, LLVMGetElementType (t)), const_int32 (0), "");
```
Comment (EgorBo) [inline src/mono/mono/mini/mini-llvm.c]: There are 4 `ShiftRightArithmetic` overloads
```csharp
public static Vector128<short> ShiftRightArithmetic(Vector128<short> value, Vector128<short> count)
public static Vector128<int> ShiftRightArithmetic(Vector128<int> value, Vector128<int> count)
public static Vector128<short> ShiftRightArithmetic(Vector128<short> value, byte count)
public static Vector128<int> ShiftRightArithmetic(Vector128<int> value, byte count)
```
It looks like this code only handles `Vector` as a count 
(it crashes for me when I try to pass a byte variable)

Comment (vargaz) [reply src/mono/mono/mini/mini-llvm.c]: There are a bunch of methods/overloads with missing tests.

**Comment by EgorBo** [comment]
> Disable SSE2 support as some intrinsics are still missing.

Looks like the following are missing:
```
ConvertScalarToVector128Double
ConvertScalarToVector128Single
ShiftLeftLogical
ShiftRightLogical
```

**Comment by EgorBo** [comment]
@tannergooding do you happen to know how coreclr handles intrinsics which require constant args and a variable is passed instead? e.g.
```csharp
Vector128<int> Maaa(Vector128<int> lhs, byte rhs)
{
    return Sse2.Shuffle(lhs, rhs);
}
```

**Review by EgorBo — APPROVED** [review]

**Comment by tannergooding** [comment]
> do you happen to know how coreclr handles intrinsics which require constant args and a variable is passed instead

If the value is constant, we just return a `gtNewSimdHWIntrinsicNode` immediately. Otherwise, we return `nullptr` which forces a standard call node to be emitted.
When it becomes time to generate code for the call, we'll see it is recursively calling itself (tracked via [mustExpand](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsic.cpp#L631)) and we'll return a `gtNewSimdHWIntrinsicNode` node in that case anyways.
This is done to support indirect calls (delegates, reflection, debugger, etc) as well as instructions that require a constant input for their operand (such as `shufps`).

When it gets to codegen, we then check if the value is `constant` again, for example see [here](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsiccodegenxarch.cpp#L239-L253).
If the value is constant, codegen can happen normally and we just emit the instruction. However, if it isn't we go to [genHWIntrinsicJumpTableFallback](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsiccodegenxarch.cpp#L1069) and emit a jump table containing all possible scenarios (up to 256) and hard code the constants in the table.

The fallback is slow, but it ensures things will work in various edge scenarios and devs using HWIntrinsics should be profiling so they'll catch the issue. We've previously discussed an analyzer as something that could help flag non-constant inputs to the user as well, but that hasn't been implemented yet. We also have an issue (https://github.com/dotnet/runtime/issues/11062) tracking us delaying the `mustExpand` decision until later (such as in lowering) so we can cover more constant input cases, but that will require some more work in the JIT.

CC. @CarolEidt, @echesakovMSFT

**Comment by EgorBo** [comment]
> However, if it isn't we go to [genHWIntrinsicJumpTableFallback](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsiccodegenxarch.cpp#L1069) and emit a jump table containing all possible scenarios (up to 256) and hard code the constants in the table.

Thanks, so it's basically the same as what Zoltan did (not sure about the 256 limit)

**Comment by tannergooding** [comment]
>  I see that all imm arguments are of byte type in C# API for that :) even if it's int in the C API

Yes, because the underlying instruction only accepts an `imm8` (hence why 256 is also the upper limit of the number of jump table scenarios).

**Pull request merged**